common/lib/common-definitions/src/logger.ts

```diff
@@ -45,12 +45,166 @@ export interface ITelemetryBaseEvent extends ITelemetryProperties {
+     */
+    id: string;
+    /*
+     * Flag indicating if call suceeded or failed.
+     */
+    success: boolean;
+    /*
+     * Optional response status.
+     */
+    status?: string;
+    /*
+     * JSON stringifed bag od details on the event.
+     */
+    details?: string;
+}
+
+/**
+ * General-use Error event. Typically represents an exception that causes an operation to fail.
+ */
+export interface ITelemetryGenErrorEvent extends ITelemetryGenEventBase {
+    type: "error";
+    /*
+     * Used for exceptions grouping (ex exception type + function).
+     */
+    errorCode: string;
+    /*
+     * Optional message.
+     */
+    message?: string;
+    /*
+     * Severity level.
+     */
+    severityLevel: TelemetryErrorCategory;
+    /*
+     * JSON stringifed stack trace
+     */
+    stackTrace?: string;
+    /*
+     * JSON stringifed details on condition(s) leading to this error
+     */
+    details?: string;
+}
+
+/**
+ * General-use Service event. Represents a call from FF to an external service or storage.
+ */
+export interface ITelemetryGenServiceEvent extends ITelemetryGenEventBase {
+    type: "service";
+    /*
+     * Duration of the call.
+     */
+    duration: number;
+    /*
+     * Identifier of a dependency call instance (correlation id).
+     */
+    id?: string;
+    /*
+     * Target site of a dependency call. Examples are server name, host address (with PII considerations).
+     */
+    target: string;
+    /*
+     * Result code. Ex. HTTP status code.
+     */
+    resultCode: string;
+    /*
+     * Flag indicating success/failure.
+     */
+    success: boolean;
+}
+
+/**
+ * General-use class event - Represents an event firing on a specific object.
+ * (Need a better name. "Event" has few different meanings in logger.)
+ */
+export interface ITelemetryGenClassEvent extends ITelemetryGenEventBase {
+    type: "event";
+    /*
+     * JSON stringifed bag od details on the event.
+     */
+    details?: string;
+}
+
+export type ITelemetryGenEvent =
+    | ITelemetryGenApiEvent
+    | ITelemetryGenServiceEvent
+    | ITelemetryGenErrorEvent
+    | ITelemetryGenClassEvent;
+
 /**
  * Interface to output telemetry events.
  * Implemented by hosting app / loader
  */
 export interface ITelemetryBaseLogger {
+    /**
+     * Unstructured dev-only telemetry. Event properties can change across minor or major releases.
+     * This stream is inteded to generate telemetry that is actionable to Fluid developers.
+     * Implemented by derived classes
+     * @param event - Telemetry event to send over
+     */
     send(event: ITelemetryBaseEvent): void;
+
+    /**
+     * General-use, static telemetry that is meant for general consumption. Events are strongly typed and follow
+     * minor/major "breaking" rules.
+     * Implemented by derived classes
+     * @param event - General-use telemetry event to send over
+     */
+    sendGenTelemetry?(event: ITelemetryGenEvent): void;
```

Comment:
i would prefer to just add a new event on IContainer that emits this stuff, rather than putting it on our based telemetry class. internally in telemetry utils you could plumb a method like this through to make emitting easier, you could also log it by default. it would also allow you to implement this at a much better layer of the stack, as the common is really about sharing between the client and sever.

Reply:
Yes, one consideration here is if we structure any telemetry it will seem symmetrical to an API surface. Then next question is, why not put it there then :) I’ll think about the approach you suggested. But, before any solution let’s consider again what we are trying to achieve here in terms of dev & client experience. If we solidify goals we may have an easier time reasoning how we get there. The rough set of "wants", mostly based on conversations with 3Ps:

(1) Consumers should deal with single, unified telemetry model, while being able to group/filter based on their needs. “needs” is a brooder topic, but think of categories, specific packages/layers they are working with etc.
(2) All of our telemetry should play nicely with each other, with no duplication of information. Our developers should not need to emit the same data across different telemetry categories.
(3) We should be able to instrument any FF package. Customers who work with FluidContainer, see the world from FluidContainer perspective and we should to be instrument the models that customer is familiar and dealing with. All FF packages are “Fluid Framework” to them.
(4) Opt-in into telemetry should be simple.
(5) We should have ability to start with smaller scope and adopt changes gradually.
(6) ?

If we add eventing on IContainer, we are effectively introducing two different telemetry concepts (see 1 above). If we are aiming for (2), the divergence of telemetry channels may add confusion how to interpret telemetry in holistic manner. (3) may also not play nice.

Either way, the preference here is to align what we want to achieve, and then let's work backwards.

Reply:
I'd suggest having a design doc, and a meeting.

If the main issue is - there is too much information in existing logs, then this is rather easy to solve (adding better articulation via category, severity / priority, etc.).
If the issue here - not clear what layer logging statement belongs too - we can solve it too in existing infrastructure.
Basically, almost always incrementally solving problems is better than trying to build a new system :) Learning from most of the rewrites - it's easy to see what's wrong with the existing system, it's very hard to notice what it does right, and usually when rewrites happen, it takes 5% of effort to fix what's wrong, and then 95% of effort is spent trying to replicate what already works great.